### PR TITLE
INFRA-835: Made the service more robust against auth_api failures.

### PR DESCRIPTION
**Pull request by yellottyellott**
This sets the user info cookie expiration to the same as the auth proxy cookie expiration.
`--auth-api-refresh` no longer sets the expiration for the user info cookie. Instead, it is compared to when the  user info cookie was written, and hits the `refresh` API if enough time has passed.

If the `auth_api` is down, logged in users won't get `refresh`ed, but will continue to use their existing user info stored in their cookie.

If a user has been deleted, both cookies will be cleared and the user will be logged out.

### CODE REVIEW

**Comment by mccutchen** [comment]
@yellottyellott this looks pretty good so far!  Think we could add tests for some of these corner cases we have now?

**Comment by yellottyellott** [comment]
Yep yep, I'll add some test cases in.

File: user_info_handler.go [hunk 1]
```diff
@@ -27,26 +29,54 @@ type UserInfo map[string]string
...
-		goto Save
-	}
-	userInfo, err = m.FetchUserInfo(ses.User)
 	if err != nil {
-		return err
+		// Valid cookie doesn't exist. Fetch info.
+		fetchUserInfo = true
+	} else {
+		// Cookie exists. Check if it needs to be refreshed.
+		c, _ := req.Cookie(m.cookieName)
```
Comment (jonmorehouse) [inline user_info_handler.go]: Should we be ignoring these errors here?

Comment (jonmorehouse) [inline user_info_handler.go]: (Not sure of the implications)

Comment (yellottyellott) [inline user_info_handler.go]: Yeah, I should expand these back out.

**Comment by yellottyellott** [comment]
To test this, I had to add an interface into `auth_api` to mock out the API request. 
I broke some stuff down in `user_info_handler` to make it easier to test. 

@jonmorehouse Could you take a look at my interface junk?

File: user_info_handler.go [hunk 2]
```diff
@@ -23,80 +24,86 @@ type UserInfoHandler struct {
...
+			if cachedInfo != nil {
+				return cachedInfo, false, nil
+			} else {
+				return nil, false, err
+			}
+		}
+		if !userExists {
+			// If the user no longer exists, clear this session.
+			err = fmt.Errorf("user no longer exists: %s", ses.User)
+			return nil, false, err
```
Comment (jonmorehouse) [inline user_info_handler.go]: Probably can just do this all in the one line

File: auth_api.go [hunk 3]
```diff
@@ -25,53 +30,66 @@ type AuthApiResponse struct {
 	Roles    []string `json:"roles"`
 }
 
-func NewAuthApiFromUrl(rawurl string) (*AuthApi, error) {
+func NewAuthApiFromUrl(rawurl string) (*DefaultAuthApi, error) {
```
Comment (mccutchen) [inline auth_api.go]: I think the type of the first return val here should be `AuthApi` instead of `*DefaultAuthApi`

**Comment by mccutchen** [comment]
@yellottyellott :heart: :heart: :heart: :heart: :heart: these tests

**Comment by mccutchen** [comment]
This LGTM assuming those beautiful new tests are passing, leaving for @jonmorehouse.  Nice work @yellottyellott!

File: auth_api.go (hunk 3 above)
Comment (yellottyellott) [inline auth_api.go]: Replacing it throws: 
```
vagrant@vagrant-ubuntu-trusty-64:/opt/go/src/github.com/buzzfeed/auth_proxy$ ./test.sh
# github.com/buzzfeed/auth_proxy
./auth_api_test.go:15: authApi.url undefined (type AuthApi has no field or method url)
./auth_api_test.go:16: authApi.url undefined (type AuthApi has no field or method url)
FAIL	github.com/buzzfeed/auth_proxy [build failed]
ok  	github.com/buzzfeed/auth_proxy/api	0.004s
ok  	github.com/buzzfeed/auth_proxy/cookie	0.003s
ok  	github.com/buzzfeed/auth_proxy/providers	0.007s
vagrant@vagrant-ubuntu-trusty-64:/opt/go/src/github.com/buzzfeed/auth_proxy$
```

Comment (mccutchen) [inline auth_api.go]: Yep, please ignore me

**Comment by yellottyellott** [comment]
Updated tests to use an anonymous struct. 
Also added some simple tests for checking the cookie age.

File: auth_api.go (hunk 3 above)
Comment (jonmorehouse) [inline auth_api.go]: Well, it'd be more idiomatic to return `AuthApi` but we can't access any attributes of the underlying type directly.

It looks like those tests are just failing because https://github.com/buzzfeed/auth_proxy/blob/master/auth_api_test.go#L16 is trying to test some internals of the underlying type.

If we think those tests are useful we can cast the interface type to the _real_ type (which is valuable, in and of itself from a testing standpoint).

```go
url := "http://test.buzzfeed.com"
authApi, err := NewAuthApiFromUrl(url)
assert.Equal(t, err, nil)

defaultAuthApi, err := authApi(DefaultAuthApi)
assert.Equal(t, defaultAuthApi.url.Scheme, "http")
assert.Equal(t, defaultAuthApi.url.Host, "test.buzzfeed.com")
```

File: user_info_handler_test.go [hunk 4]
```diff
@@ -51,7 +65,174 @@ func TestUserInfoHeaders(t *testing.T) {
 	req, _ := http.NewRequest("GET", "/", nil)
 	handler.WriteForwardedHeaders(req, userInfo)
 
-	assert.Equal(t, req.Header["X-Forwarded-Test"][0], "test")
+	assert.Equal(t, req.Header["X-Forwarded-User"][0], "test")
+}
+
+func TestIsUserInfoStale(t *testing.T) {
+	handler := &UserInfoHandler{
+		refresh:        time.Second * 30,
```
Comment (jonmorehouse) [inline user_info_handler_test.go]: I think we need to run `gofmt` on this project, maybe! I can never remember if `golang` likes spaces around operators

Comment (yellottyellott) [inline user_info_handler_test.go]: i already did. it's the last commit.

Comment (jonmorehouse) [inline user_info_handler_test.go]: Oh cool! JW

File: auth_api.go (hunk 3 above)
Comment (yellottyellott) [inline auth_api.go]: Should we just define `.URL()` as part of the interface?
I'm guessing we can't specify an attribute as part of the interface.

Comment (yellottyellott) [inline auth_api.go]: Or maybe this func should be named `NewDefaultAuthApiFromUrl`?

Comment (mccutchen) [inline auth_api.go]: I feel like it's fine as-is, unless @jonmorehouse feels strongly

Comment (jonmorehouse) [inline auth_api.go]: I don't feel overly strongly; +1 to whatever is easiest!

I think we should leave the interface as is. I was just proposing modifying tests to get things working. Ideally, we want the minimal surface area for the interface if possible.

**Comment by yellottyellott** [comment]
Ready to :hammer: ?

**Comment by mccutchen** [comment]
:hammer:

**Comment by yellottyellott** [comment]
:hammer: 'd

**Comment by jonmorehouse** [comment]
Nice work @yellottyellott

**Comment by mccutchen** [comment]
@yellottyellott let's get this new version into chef and get it deployed so we don't surprise ourselves later on if it turns out that there are any issues down the road!
